doc/csharp/csharp-null-ref-types-proposal.md

```diff
@@ -0,0 +1,197 @@
+a non-null message or stream must be passed to a service, and a non-null value 
+is returned.
+
+However this can break existing user code. Adding these attributes to the base 
+class may cause errors (not warnings) in user code that has already been derived 
+from the base class:
+
+> *error CS8765: Nullability of type of parameter 'request' doesn't match 
+overridden member (possibly because of nullability attributes).*
+
+Note: the advantages of adding these attributes to the service methods is 
+questionable as the only caller of these methods is the server hosting the 
+services, not user code. Therefore this may not be an issue if it is not needed 
+to be done.
+
+### 5. Keeping the generated code simple and consistent
+
+It would be possible to generate different code for nullable and non-nullable contexts
+depending on some configuration, or to generate conditional code containing `#if` 
+everywhere.
+
+Generating different code depending on configuration could lead to confusion or errors
+if code generated in one project with nullable enabled was consumed in another project 
+that did not have nullable enabled. It would also have an impact on the level of 
+testing that would need to be done.
+
+Generating conditional code would produce messy and potentially convoluted code, again
+having an impact on testing.
+
+## Disadvantages of this propsal
+
+There are a few places where adding annotations or attributes to the generated
+code to help with the static flow analysis could be useful. These are:
+- Checking non-null setters for certain fields in a message at compile time,
+ such as fields that are strings. Runtime checks are still done in the
+ generated code.
+- Checking a non-null message or stream is passed to a gRPC service call.
+- Specifying a non-null message or stream is returned from a gRPC service call.
+
+By not supporting nullable reference types in the generated code these compile
+time checks cannot be done.
+
+## Outside the scope of this proposal
+
+### Changes to related packages
+
+Changes to `Google.Protobuf` library and other libraries 
+(such as `Google.Api.CommonProtos`) are outside the scope of this proposal.
+
+There is a separate long-standing effort to enable null reference types for 
+`Google.Protobuf`.
+
+### Exposing protocol buffers optional fields as nullable properties
+
+See https://github.com/protocolbuffers/protobuf/issues/9083
+
+This has been rejected by the protocol buffers team.
+
+> "We get requests in some other languages that have null support to allow 
+optional fields to have nullable properties. The main problem is that we like to 
+keep languages consistent and other languages don't have those same options. 
+Additionally, it's not clear with some primitive fields what a null value would 
+be (is it the default? What if users meant to set it to that option?). For those 
+reasons, we will not support this in the near future.”
```

Comment:
Nit: opening quote is straight; closing quote is curly. Perhaps remove them entirely given that you're quoting anyway?